boards/p1am_100/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+        aliases: {
+            AlternateC: Scl
+        }
+    }
+    PA10 {
+        name: d2,
+        aliases: {
+            AlternateG: I2sSerialClock
+        }
+    }
+    PA11 { name: d3 }
+    PA12 {
+        name: pa12,
+        aliases: {
+            AlternateC: SdSdo
+        }
+    }
+    PA13 {
+        name: pa13,
+        aliases: {
+            AlternateC: SdSck
+        }
+    }
+    PA14 {
+        name: pa14,
+        aliases: {
+            PushPullOutput: SdSlaveSelect
+        }
+    }
+    PA15 {
+        name: pa15,
+        aliases: {
+            AlternateC: SdSdi
+        }
+    }
+    PA16 {
+        name: d8,
+        aliases: {
+            AlternateC: Spi0Mosi
+        }
+    }
+    PA17 {
+        name: d9,
+        aliases: {
+            AlternateC: Spi0Sck
+        }
+    }
+    PA18 {
+        name: pa18,
+        aliases: {
+            /// Host Enable, drive high to switch into USB host mode
+            #[cfg(feature = "usb")]
+            PushPullOutput: UsbId
+        }
+    }
+    PA19 {
+        name: d10,
+        aliases: {
+            AlternateC: Spi0Miso
+        }
+    }
+    PA20 { name: d6 }
+    PA21 { name: d7 }
+    PA22 { name: d0 }
+    PA23 { name: d1 }
+    PA24 {
+        name: usb_dm,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDm
+        }
+    }
+    PA25 {
+        name: usb_dp,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDp
+        }
+    }
+    PA27 {
+        name: pa27,
+        aliases: {
+            PullUpInput: SdCardDetect
+        }
+    }
+    PA28 {
+        name: switch,
+        aliases: {
+            PullUpInput: Switch
+        }
+    }
+    PB02 {
+        name: d16,
+        aliases: {
+            AlternateB: A1
+        }
+    }
+    PB03 {
+        name: d17,
+        aliases: {
+            AlternateB: A2
+        }
+    }
+    PB08 {
+        name: led,
+        aliases: {
+            PushPullOutput: Led
+        }
+    }
+    PB09 {
+        name: pb09,
+        aliases: {
+            PushPullOutput: BaseEnable,
+            AlternateB: AdcBattery
+        }
+    }
+    PB10 { name: d4 }
+    PB11 { name: d5 }
+    PB22 {
+        name: d14,
+        aliases: {
+            AlternateD: UartTx
+        }
+    }
+    PB23 {
+        name: d13,
+        aliases: {
+            AlternateD: UartRx
+        }
+    }
+);
+
+const BASE_CONTROLLER_FREQ: Hertz = Hertz(1000000);
+// FIXME: const BASE_CONTROLLER_SPI_MODE: dyn spi::Mode = &spi::MODE_2;
+
+/// Convenience for setting up the labeled SPI0 peripheral.
+/// SPI0 has the P1AM base controller connected.
+/// This powers up SERCOM1 and configures it for talking to the
+/// base controller.
+pub fn base_controller_spi(
+    clocks: &mut GenericClockController,
+    sercom1: pac::SERCOM1,
+    pm: &mut pac::PM,
+    sck: Spi0Sck,
+    mosi: Spi0Mosi,
+    miso: Spi0Miso,
+) -> sercom::v2::spi::Spi<impl spi::ValidConfig> {
```

Comment:
A similar argument applies here. By using `impl ValidConfig`, you hide the exact `Pads`, `Mode` and `CharSize` from the caller. That's probably fine for most cases, but if you ever wanted to store the SPI struct in a `static` variable, you would be **required** to fully name the type. Just FYI.

I understand why you did it this way though. The `Pads` type in the current version of `v2::spi` is overly complex. I provided the `spi_pads_alias` macro to help with that. But I later realized that I could refactor the code to make the `Pads` struct look almost exactly like the macro. In the latest (unpublished) version, I updated the `Pads` struct and got rid of the macro.

Reply:
I know you say you prefer letting the compiler infer the return types, but I think you'll discover that approach doesn't work well in this particular instance.

The `Spi` type is made to be used with the various `embedded_hal` traits. But those traits are only implemented for specific `Config` and `Pads` types. The idea is to use the type system to prevent misconfiguration.

In this case, though, you've only told the caller that the `spi::Config` type satisfies `ValidConfig`, nothing more. So the caller will have no idea what the `Spi` struct is capable of, and won't be able to use any of the `embedded_hal` traits.

If you want to use an existential type for this, I would probably choose `impl FullDuplex` or something like that. As it stands right now, I don't think you can do anything useful at all with the return value of this function.